akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala

```diff
@@ -922,6 +934,55 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
       }
     }
 
+    "enforce settings" should {
+
+      "reject new substreams when exceeding SETTINGS_MAX_CONCURRENT_STREAMS" in new TestSetup with RequestResponseProbes {
+        def maxStreams: Int = 16
+        override def settings: ServerSettings = super.settings.mapHttp2Settings(_.withMaxConcurrentStreams(maxStreams))
+        val requestHeaderBlock: ByteString = HPackSpecExamples.C41FirstRequestWithHuffman
+
+        // start as many streams as max concurrent...
+        private val streamIds: IndexedSeq[Int] = (0 until maxStreams).map(id => 1 + id * 2)
+        streamIds.foreach(streamId =>
+          sendHEADERS(streamId, endStream = false, endHeaders = true, requestHeaderBlock)
+        )
+
+        // while we don't exceed the limit, there's silence on the line
+        expectNoBytes()
+
+        // When we exceed the limit, though...
+        val lastValidStreamId = streamIds.max
+        val firstInvalidStreamId = lastValidStreamId + 2
+        sendHEADERS(firstInvalidStreamId, endStream = false, endHeaders = true, requestHeaderBlock)
+        val (_, code) = expectGOAWAY(lastValidStreamId)
+        code should ===(ErrorCode.PROTOCOL_ERROR)
+      }
+
+      "reject new substreams when exceeding SETTINGS_MAX_CONCURRENT_STREAMS (with closed streams in between)" in new TestSetup with RequestResponseProbes with Http2FrameHpackSupport {
+        def maxStreams: Int = 32
+        val requestHeaderBlock: ByteString = HPackSpecExamples.C41FirstRequestWithHuffman
+        override def settings = super.settings.mapHttp2Settings(_.withMaxConcurrentStreams(maxStreams))
+
+        // the Seq of stream ids has gaps
```

Comment:
If the implementation becomes more strict in the future it might not allow gaps and require you to actually open and close those streams - but OK to deal with that then ;)

Reply:
Is there any indication that they would change this?

It seems like a conscious feature. The way it is currently used, e.g. by firefox to set up a priority hierarchy with the first few streams but never actually use those streams. So, those streams will be immediately closed but if you track priorities correctly, you will use those stream identifiers as part of the priority hierarchy.

Reply:
Interesting! I thought the spec might disallow gaps but indeed it doesn't (and that won't change)